code/framework/extension/src/main/java/io/cattle/platform/extension/impl/ExtensionMap.java

```diff
@@ -0,0 +1,81 @@
+package io.cattle.platform.extension.impl;
+
+import io.cattle.platform.extension.ExtensionManager;
+
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class ExtensionMap<K, V> implements Map<K, V> {
+
+    Map<K, V> map;
+    ConcurrentHashMap<K, V> inner;
+    String key;
+    ExtensionManager extensionManager;
+
+    public ExtensionMap(ExtensionManager extensionManager, String key, Map<K, V> map) {
+        this.extensionManager = extensionManager;
+        this.key = key;
+        this.inner = map == null ? new ConcurrentHashMap<K, V>() : new ConcurrentHashMap<>(map);
+        this.map = Collections.unmodifiableMap(inner);
```

Comment:
Can you explain why we need an unmodifiable map here? Seems overly complex, especially since you hold onto a modifiable version of the map (inner) and modify it when needed.